x/gasless/types/msgs.go

```diff
@@ -0,0 +1,375 @@
+func NewMsgUpdateGasTankConfig(
+	gasTankID uint64,
+	provider sdk.AccAddress,
+	maxFeeUsagePerTx sdkmath.Int,
+	maxTxsCountPerConsumer uint64,
+	maxFeeUsagePerConsumer sdkmath.Int,
+	txsAllowed []string,
+	contractsAllowed []string,
+) *MsgUpdateGasTankConfig {
+	return &MsgUpdateGasTankConfig{
+		GasTankId:              gasTankID,
+		Provider:               provider.String(),
+		MaxFeeUsagePerTx:       maxFeeUsagePerTx,
+		MaxTxsCountPerConsumer: maxTxsCountPerConsumer,
+		MaxFeeUsagePerConsumer: maxFeeUsagePerConsumer,
+		TxsAllowed:             txsAllowed,
+		ContractsAllowed:       contractsAllowed,
+	}
+}
+
+func (msg MsgUpdateGasTankConfig) Route() string { return RouterKey }
+
+func (msg MsgUpdateGasTankConfig) Type() string { return TypeMsgUpdateGasTankConfig }
+
+func (msg MsgUpdateGasTankConfig) ValidateBasic() error {
+	if _, err := sdk.AccAddressFromBech32(msg.Provider); err != nil {
+		return sdkerrors.Wrapf(errors.ErrInvalidAddress, "invalid provider address: %v", err)
+	}
+	if msg.GasTankId == 0 {
+		return sdkerrors.Wrap(errors.ErrInvalidRequest, "gas tank id must not be 0")
+	}
+	if msg.MaxTxsCountPerConsumer == 0 {
+		return sdkerrors.Wrap(errors.ErrInvalidRequest, "max tx count per consumer must not be 0")
+	}
+	if !msg.MaxFeeUsagePerTx.IsPositive() {
+		return sdkerrors.Wrap(errors.ErrInvalidRequest, "max_fee_usage_per_tx should be positive")
+	}
+	if !msg.MaxFeeUsagePerConsumer.IsPositive() {
+		return sdkerrors.Wrap(errors.ErrInvalidRequest, "max_fee_usage_per_consumer should be positive")
+	}
+	if len(msg.TxsAllowed) == 0 && len(msg.ContractsAllowed) == 0 {
+		return sdkerrors.Wrap(errors.ErrInvalidRequest, "atleast one tx or contract is required to initialize")
+	}
+	return nil
+}
+
+func (msg MsgUpdateGasTankConfig) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&msg))
+}
+
+func (msg MsgUpdateGasTankConfig) GetSigners() []sdk.AccAddress {
+	addr, err := sdk.AccAddressFromBech32(msg.Provider)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{addr}
+}
+
+// NewMsgBlockConsumer returns a new MsgBlockConsumer.
+func NewMsgBlockConsumer(
+	gasTankID uint64,
+	actor, consumer sdk.AccAddress,
+) *MsgBlockConsumer {
+	return &MsgBlockConsumer{
+		GasTankId: gasTankID,
+		Actor:     actor.String(),
+		Consumer:  consumer.String(),
+	}
+}
+
+func (msg MsgBlockConsumer) Route() string { return RouterKey }
+
+func (msg MsgBlockConsumer) Type() string { return TypeMsgBlockConsumer }
+
+func (msg MsgBlockConsumer) ValidateBasic() error {
+	if msg.GasTankId == 0 {
+		return sdkerrors.Wrap(errors.ErrInvalidRequest, "gas tank id must not be 0")
+	}
+	if _, err := sdk.AccAddressFromBech32(msg.Actor); err != nil {
+		return sdkerrors.Wrapf(errors.ErrInvalidAddress, "invalid actor address: %v", err)
+	}
+	if _, err := sdk.AccAddressFromBech32(msg.Consumer); err != nil {
+		return sdkerrors.Wrapf(errors.ErrInvalidAddress, "invalid consumer address: %v", err)
+	}
```

Comment:
this particular block was repeated a few times, please consider moving it to a new method, in order to promote DRYness